staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -135,6 +137,130 @@ func (e *watchEmbeddedEncoder) embeddedIdentifier() runtime.Identifier {
+	defer e.buffer.Reset()
+
+	if err := e.embeddedEncoder.Encode(obj, e.buffer); err != nil {
+		return fmt.Errorf("unable to encode watch object %T: %v", obj, err)
+	}
+
+	// ContentType is not required here because we are defaulting to the serializer type.
+	e.unknown.Raw = e.buffer.Bytes()
+	event.Object = &e.unknown
+	defer func() {
+		// Restore the original event to not modify the object as part of Encode.
+		event.Object = obj
+	}()
+	metrics.WatchEventsSizes.WithContext(e.ctx).WithLabelValues(e.kind.Group, e.kind.Version, e.kind.Kind).Observe(float64(len(e.unknown.Raw)))
+
+	*e.outEvent = metav1.WatchEvent{}
+
+	// create the external type directly and encode it.  Clients will only recognize the serialization we provide.
+	// The internal event is being reused, not reallocated so its just a few extra assignments to do it this way
+	// and we get the benefit of using conversion functions which already have to stay in sync
+	*e.internalEvent = metav1.InternalEvent(event)
+	if err := metav1.Convert_v1_InternalEvent_To_v1_WatchEvent(e.internalEvent, e.outEvent, nil); err != nil {
+		return fmt.Errorf("unable to convert watch object: %v", err)
+	}
+
+	defer e.eventBuffer.Reset()
+	if err := e.encoder.Encode(e.outEvent, e.eventBuffer); err != nil {
+		return fmt.Errorf("unable to encode watch object %T: %v (%#v)", e.outEvent, err, e)
+	}
+
+	_, err := w.Write(e.eventBuffer.Bytes())
+	return err
+}
+
+type watchEncoderIdentifier struct {
+	Name            string `json:"name,omitempty"`
+	EmbeddedEncoder string `json:"embeddedEncoder,omitempty"`
+	Encoder         string `json:"encoder,omitempty"`
+	EventType       string `json:"eventType,omitempty"`
+}
+
+func (e *watchEncoder) identifier(eventType watch.EventType) runtime.Identifier {
+	// We need to take into account that in embeddedEncoder includes table
+	// transformer, then its identifier is dynamic. As a result, whenever
+	// the identifier of embeddedEncoder changes, we need to invalidate the
+	// whole identifiers cache.
+	// TODO(wojtek-t): Can we optimize it somehow?
+	if e.currentEmbeddedIdentifier != e.embeddedEncoder.Identifier() {
+		e.currentEmbeddedIdentifier = e.embeddedEncoder.Identifier()
+		e.identifiers = map[watch.EventType]runtime.Identifier{}
+	}
+	if _, ok := e.identifiers[eventType]; !ok {
+		e.identifiers[eventType] = e.typeIdentifier(eventType)
+	}
+	return e.identifiers[eventType]
+}
+
+func (e *watchEncoder) typeIdentifier(eventType watch.EventType) runtime.Identifier {
+	// The eventType is a non-standard pattern. This is coming from the fact
+	// that we're effectively serializing the whole watch event, but storing
+	// it in serializations of the Object within the watch event.
+	identifier := watchEncoderIdentifier{
+		Name:            "watch",
+		EmbeddedEncoder: string(e.embeddedEncoder.Identifier()),
+		Encoder:         string(e.encoder.Identifier()),
+		EventType:       string(eventType),
```

Comment:
This is combining multiple encoder identities (other places tend to wrap a single nested encoder). Do you have an example of what this ends up being in practice, or how big? (I've lost track of how many nested levels we've gotten at this point)

It also has to include the entire content of the watch event (other than the object itself) as the cache identifier ... that happens to just be the type field, but if other fields were added to watch event, they'd have to be added here as part of the key as well, right? That's the part that struck me as twisty.

Reply:
> This is combining multiple encoder identities (other places tend to wrap a single nested encoder). Do you have an example of what this ends up being in practice, or how big? (I've lost track of how many nested levels we've gotten at this point)

It has quite a few already, e.g.:

```
"{\"name\":\"watch\",\"embeddedEncoder\":\"{\\\"encodeGV\\\":\\\"v1\\\",\\\"encoder\\\":\\\"protobuf\\\",\\\"name\\\":\\\"versioning\\\"}\",\"encoder\":\"{\\\"encodeGV\\\":\\\"v1\\\",\\\"encoder\\\":\\\"raw-protobuf\\\",\\\"name\\\":\\\"versioning\\\"}\",\"eventType\":\"MODIFIED\"}
```


Regarding adding new fields to watch event - I added a trivial unit test that checks that with a error message that captures what should be done in case of failure.